r2dbc-spi/src/main/java/io/r2dbc/spi/OutParametersMetadata.java

```diff
@@ -0,0 +1,71 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.spi;
+
+import java.util.Collection;
+import java.util.NoSuchElementException;
+
+/**
+ * Represents the metadata for {@code OUT} parameters of the results returned from a stored procedure.
+ * Metadata for parameters can be either retrieved by specifying a parameter name or the parameter index.
+ * Parameter indexes are {@code 0}-based.
+ */
+public interface OutParametersMetadata {
+
+    /**
+     * Returns the {@link OutParameterMetadata} for one parameter.
+     *
+     * @param index the parameter index starting at 0
+     * @return the {@link OutParameterMetadata} for one parameter
+     * @throws ArrayIndexOutOfBoundsException if the {@code index} is less than zero or greater than the number of available parameters.
+     */
+    OutParameterMetadata getParameterMetadata(int index);
+
+    /**
+     * Returns the {@link OutParameterMetadata} for one parameter.
+     *
+     * @param name the name of the parameter.  Parameter names are case insensitive.  When a get method contains several parameters with same name, then the value of the first matching parameter
+     *             will be returned
+     * @return the {@link OutParameterMetadata} for one parameter
+     * @throws IllegalArgumentException if {@code name} is {@code null}
+     * @throws NoSuchElementException   if there is no parameter with the {@code name}
+     */
+    OutParameterMetadata getParameterMetadata(String name);
+
+    /**
+     * Returns the {@link OutParameterMetadata} for all parameters.
+     *
+     * @return the {@link OutParameterMetadata} for all parameters
+     */
+    Iterable<? extends OutParameterMetadata> getParameterMetadatas();
+
+    /**
+     * Returns an unmodifiable collection of parameter names.
+     * <p>
+     * Any attempts to modify the returned collection, whether direct or via its iterator, result in an {@link UnsupportedOperationException}.
+     * <p>
+     * The iteration order of the parameter names depends on the actual stored procedure.
+     * Parameter names may appear multiple times if the result specifies multiple parameters with the same name.
+     * Lookups through {@link Collection#contains(Object)} are case-insensitive.
+     * Drivers may enhance comparison sorting rules with escape characters to enforce a particular mode of comparison
+     * when querying for presence/absence of a parameter.
+     *
+     * @return the parameter names.
+     */
+    Collection<String> getParameterNames();
+
+}
```

Comment:
I have a few concerns regarding this type here. It mimicks the API of `RowMetadata` as it appears to be a similar thing at first. Yet it is very different in various ways.

1. Unlike top-level `SELECT` statements that allow duplicate column names, as far as I am aware, there aren't any dialects that allow duplicate parameter names in a procedure. Perhaps there is no need to mention a hypothetical edge case in `getParameterMetadata(String)`'s Javadoc?
2. If designed as it is right now, the indexes from `getParameterMetadata(int)` will shift from the original indexes when preparing the call. Consider:

    ```sql
    procedure p(p0 in out int, p1 in int, p2 out int);
    ```

    When registering/binding parameters, the indexes are: `0: p0`, `1: p1`, `2: p2`, but when looking at `OUT` parameters exclusively, we now get `0: p0`, `1: p2`. This is very confusing and error prone. Leading to the next point:
3. The above can be either accepted as a necessary evil of the current design, or the `OutParametersMetadata` type is changed to `ParametersMetadata`, including meta data for `IN` parameters. I don't see why not.

Reply:
Addressing 1. makes sense as we should expect unique parameter names.

Stored procedures indeed lead to increased mental overhead. Retrieving IN parameters through `row.get(…)` seems a bit weird and in JDBC-land that doesn't seem to work either. Throwing an exception if an `IN` parameter is requested is probably the only sensible thing to do if we do not want to mess with indexes. 

On the other hand, limiting parameter metadata to out parameters only allows contiguous by-index retrieval. That simplifies the entire arrangement a bit. Since R2DBC can make use of proper parameter names, where supported, giving guidance towards a named parameters preference would help to get away from a by-index mental model.

What also plays in this abstraction is the retrieval of return values that isn't an explicit OUT parameter registration-wise but it can be returned by databases. 

That being said, I'm in favor of keeping the scope limited to things that a stored procedure produces. Happy to continue the discussion on a Friday's R2DBC call.

On a related note, as outcome of #219, I would also remove `Collection<String> getParameterNames()` (without replacement).

Reply:
> giving guidance towards a named parameters preference would help to get away from a by-index mental model.

Here's my obligatory objection from a performance perspective. Access by index should always be preferred by clients, if possible. Quick ad-hoc procedure calls can use names, sure. But tools built on top of the API should access by index.

> What also plays in this abstraction is the retrieval of return values that isn't an explicit OUT parameter registration-wise but it can be returned by databases.

Can any database do that? In PL/SQL, you cannot. You have to explicitly assign *each* OUT parameter to a local variable. You can't just ignore any results. You can't even ignore function results.

> Happy to continue the discussion on a Friday's R2DBC call.

I can't make it, I'm afraid.

> On a related note, as outcome of #219, I would also remove `Collection<String> getParameterNames()` (without replacement).

👌

Reply:
> Here's my obligatory objection from a performance perspective…

Agree with your arguments.

>> What also plays in this abstraction is the retrieval of return values that isn't an explicit OUT parameter registration->> wise but it can be returned by databases.
> Can any database do that?

Good question. I've seen it for SQL server only so far. If there's no other database then the SQL server driver can come up with something specific to their needs without polluting the spec.